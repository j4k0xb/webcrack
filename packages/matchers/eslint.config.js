import config from '@webcrack/eslint-config';

/**
 * @type {import('eslint').Linter.Config[]}
 */
export default [
  ...config,
  {
    ignores: ['src/generated.ts'],
  },
];
