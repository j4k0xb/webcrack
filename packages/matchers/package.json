{
  "name": "@webcrack/matchers",
  "version": "0.1.0",
  "private": true,
  "type": "module",
  "main": "src/index.ts",
  "scripts": {
    "build": "tsx scripts/build.ts",
    "bench": "tsx benchmark/test.js",
    "lint": "eslint src scripts",
    "lint:fix": "eslint src scripts --fix",
    "format:check": "prettier --check \"{src,scripts}/**/*.ts\"",
    "format": "prettier --write \"{src,scripts}**/*.ts\"",
    "typecheck": "tsc --noEmit",
    "test": "vitest --no-isolate"
  },
  "dependencies": {
    "@babel/parser": "^7.27.0",
    "@babel/types": "^7.27.0"
  },
  "devDependencies": {
    "@babel/traverse": "^7.27.0",
    "@codemod/matchers": "^1.7.1",
    "@types/babel__traverse": "^7.20.7",
    "@types/node": "^22.13.16",
    "@webcrack/eslint-config": "workspace:*",
    "@webcrack/typescript-config": "workspace:*",
    "benny": "^3.7.1",
    "tsx": "^4.19.3",
    "typescript": "^5.8.2"
  }
}
