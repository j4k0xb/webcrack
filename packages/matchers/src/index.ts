export * from './compiler.js';
export * from './generated.js';
export * from './template';
export * from './types.js';
