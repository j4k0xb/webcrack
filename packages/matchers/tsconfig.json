{
  "extends": "@webcrack/typescript-config/base.json",
  "compilerOptions": {
    "noEmit": true,
    "verbatimModuleSyntax": true,
  },
  "include": [
    "*.ts",
    "src",
    "test",
    "scripts",
  ]
}
