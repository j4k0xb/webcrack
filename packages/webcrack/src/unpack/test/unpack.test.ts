import * as m from '@codemod/matchers';
import { readFile } from 'fs/promises';
import { tmpdir } from 'os';
import { join, sep } from 'path';
import { expect, test } from 'vitest';
import { webcrack } from '../../index';

const SAMPLES_DIR = join(__dirname, 'samples');

test('detect top-level bundle first', async () => {
  const { bundle } = await webcrack(
    await readFile(join(SAMPLES_DIR, 'browserify-webpack-nested.js'), 'utf8'),
  );
  expect(bundle).toBeDefined();
  expect(bundle!.type).toBe('browserify');
});

test('path mapping', async () => {
  const { bundle } = await webcrack(
    await readFile(join(SAMPLES_DIR, 'webpack-4.js'), 'utf8'),
    {
      mappings: () => ({
        './utils/color.js': m.stringLiteral('#FBC02D'),
        package: m.numericLiteral(4),
      }),
    },
  );
  expect(bundle).toBeDefined();
  expect(bundle!).toMatchSnapshot();
});

test.runIf(sep === '/')('prevent path traversal (posix)', async () => {
  const code = await readFile(
    join(SAMPLES_DIR, 'webpack-path-traversal.js'),
    'utf8',
  );
  const { bundle } = await webcrack(code);
  expect(bundle).toBeDefined();

  const dir = join(tmpdir(), 'path-traversal-test');
  await expect(bundle!.save(dir)).rejects.toThrow('path traversal');
});

test.runIf(sep === '\\')('prevent path traversal (windows)', async () => {
  const code = await readFile(
    join(SAMPLES_DIR, 'webpack-path-traversal-windows.js'),
    'utf8',
  );
  const { bundle } = await webcrack(code);
  expect(bundle).toBeDefined();

  const dir = join(tmpdir(), 'path-traversal-test');
  await expect(bundle!.save(dir)).rejects.toThrow('path traversal');
});
