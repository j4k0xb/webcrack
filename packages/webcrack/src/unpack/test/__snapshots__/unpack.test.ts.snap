// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`path mapping 1`] = `
WebpackBundle {
  "entryId": "2",
  "modules": Map {
    "1" => WebpackModule {
      "ast": const a = require("package");,
      "id": "1",
      "isEntry": false,
      "path": "./1.js",
    },
    "2" => WebpackModule {
      "ast": const a = require("../1.js");
const _module = 1;
module.exports.color = "#FBC02D";
{
  const _module2 = 2;
  console.log(_module2);
  console.log(module);
}
exports.a = 3;,
      "id": "2",
      "isEntry": true,
      "path": "./utils/color.js",
    },
    "3" => WebpackModule {
      "ast": module.exports = 4;,
      "id": "3",
      "isEntry": false,
      "path": "node_modules/package",
    },
  },
  "type": "webpack",
}
`;
